internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -825,3 +858,86 @@ func UpdateResourceRefs(xr resource.ComposedResourcesReferencer, desired Compose
+type PatchingManagedFieldsUpgrader struct {
+	client client.Writer
+}
+
+// NewPatchingManagedFieldsUpgrader returns a ManagedFieldsUpgrader that uses a
+// JSON patch to upgrade and object's managed fields from client-side to
+// server-side apply.
+func NewPatchingManagedFieldsUpgrader(w client.Writer) *PatchingManagedFieldsUpgrader {
+	return &PatchingManagedFieldsUpgrader{client: w}
+}
+
+// Upgrade the supplied composed object's field managers from client-side to server-side
+// apply.
+//
+// This is a multi-step process.
+//
+// Step 1: All fields are owned by manager 'crossplane' operation 'Update'. This
+// represents all fields set by the XR controller up to this point.
+//
+// Step 2: Upgrade is called for the first time. We clear all field managers.
+//
+// Step 3: The XR controller server-side applies its fully specified intent
+// as field manager with prefix 'apiextensions.crossplane.io/composed/'. This becomes the
+// manager of all the fields that are part of the XR controller's fully
+// specified intent. All existing fields the XR ontroller didn't specify
+// become owned by a special manager - 'before-first-apply', operation 'Update'.
+//
+// Step 4: Upgrade is called for the second time. It deletes the
+// 'before-first-apply' field manager entry. Only the XR composed field manager
+// remains.
+func (u *PatchingManagedFieldsUpgrader) Upgrade(ctx context.Context, obj client.Object) error {
+	// The composed resource doesn't exist, nothing to upgrade.
+	if !meta.WasCreated(obj) {
+		return nil
+	}
+
+	foundSSA := false
+	foundBFA := false
+	idxBFA := -1
+
+	for i, e := range obj.GetManagedFields() {
+		if strings.HasPrefix(e.Manager, FieldOwnerComposedPrefix) {
+			foundSSA = true
+		}
+		if e.Manager == "before-first-apply" {
+			foundBFA = true
+			idxBFA = i
+		}
+	}
+
+	switch {
+	// If our SSA field manager exists and the before-first-apply field manager
+	// doesn't, we've already done the upgrade. Don't do it again.
+	case foundSSA && !foundBFA:
+		return nil
+
+	// We found our SSA field manager but also before-first-apply. It should now
+	// be safe to delete before-first-apply.
+	case foundSSA && foundBFA:
+		p := []byte(fmt.Sprintf(`[
+			{"op": "remove", "path": "/metadata/managedFields/%d"},
+			{"op": "replace", "path": "/metadata/resourceVersion", "value": "%s"}
+		]`, idxBFA, obj.GetResourceVersion()))
+		return errors.Wrap(resource.IgnoreNotFound(u.client.Patch(ctx, obj, client.RawPatch(types.JSONPatchType, p))), "cannot remove before-first-apply from field managers")
+
+	// We didn't find our SSA field manager or the before-first-apply field
+	// manager. This means we haven't started the upgrade. The first thing we
+	// want to do is clear all managed fields. After we do this we'll let our
+	// SSA field manager apply the fields it cares about. The result will be
+	// that our SSA field manager shares ownership with a new manager named
+	// 'before-first-apply'.
+	default:
```

Comment:
is it possible that we could reach this `default` case without finding our SSA field manager but yes we did find the before-first-apply manager? Is that valid, and if so, this operation is OK for that case too?

Reply:
>  if so, this operation is OK for that case too?

yes, it is OK/safe to clear all managedField entries, even if `foundBFA && !foundSSA`